File: SEMANA I/introductorio_2017/kotlin/gcd/gcd.kt
/*
 Compilar
 kotlinc gcd.kt
 Run (por ejemplo gcd(235, 75)
 kotlin GcdKt 940 600
*/

/* if ( a == b ) return a;
  if ( a == 0 ) return b;
  if ( b == 0 ) return a;
  return if ( a > b ) gcd(a - b, b) 
         else gcd(a, b - a); */
		 
// Kotlin recursivo por reglas
fun gcd( a: Int, b: Int) : Int =
 when {
   a == b -> a
   a == 0 -> b
   b == 0 -> a
   a > b  -> gcd(a - b, b)
   else   -> gcd(a, b - a) 
  }



fun main(args : Array<String>){
  var a : Int = 0;
  var b : Int = 0;
  if ( args.count() >= 1){
    a = args[0].toInt();
  }
  if ( args.count() >= 2){
    b = args[1].toInt();
  }
  println("gcd($a, $b)= ${gcd(a, b)}")
  
}

File: SEMANA I/introductorio_2017/scala/gcd/gcd.scala

object Gcd {

   def gcd(a: Int, b: Int) : Int = 
     
    ( a, b) match {
	    case (0, b) => b
		case (a, 0) => a
		case (a, b) if (a > b) 
		            => gcd(a - b, b)
		case _      => gcd(a, b - a)
	  
	}
    

   def main(args: Array[String]) = {
        var a = 0; var b = 0;
		if ( args.length >= 1  ){
			a = args(0).toInt;
		}
		if ( args.length >= 2  ){
			b = args(1).toInt;
		}
		println(s"gcd($a, $b)= ${gcd(a, b)}");
		
   }
}
